net/host.go

```diff
@@ -81,25 +80,29 @@ type Config struct {
 // The host implemented in this package is swap-specific; ie. it supports swap-specific
 // messages (initiate and query).
 func NewHost(cfg *Config) (*Host, error) {
-	h, err := p2pnet.NewHost(&p2pnet.Config{
-		Ctx:        cfg.Ctx,
-		DataDir:    cfg.DataDir,
-		Port:       cfg.Port,
-		KeyFile:    cfg.KeyFile,
-		Bootnodes:  cfg.Bootnodes,
-		ProtocolID: cfg.ProtocolID,
-		ListenIP:   cfg.ListenIP,
+	h := &Host{
+		ctx:       cfg.Ctx,
+		h:         nil, // set below
+		isRelayer: cfg.IsRelayer,
+		swaps:     make(map[types.Hash]*swap),
+	}
+
+	var err error
+	h.h, err = p2pnet.NewHost(&p2pnet.Config{
+		Ctx:                      cfg.Ctx,
+		DataDir:                  cfg.DataDir,
+		Port:                     cfg.Port,
+		KeyFile:                  cfg.KeyFile,
+		Bootnodes:                cfg.Bootnodes,
+		ProtocolID:               cfg.ProtocolID,
+		ListenIP:                 cfg.ListenIP,
+		AdvertisedNamespacesFunc: h.advertisedNamespaces,
```

Comment:
The self reference on this line is why I had to reverse the initialization order.